### PR TITLE
Fix couple of errors in checklink()

### PR DESCRIPTION
**Pull request by monich**
1. Removed useless check for NULL after the pointer has already been used
2. Correctly NULL-terminate buffer filled by `readlink()`

The latter was was causing valgrind warnings like this:
```
==32357== Conditional jump or move depends on uninitialised value(s)
==32357==    at 0x484456C: strcmp (vg_replace_strmem.c:755)
==32357==    by 0x10E717: checklink (usb_moded-network.c:309)
==32357==    by 0x10E717: write_udhcpd_conf (usb_moded-network.c:395)
==32357==    by 0x10E717: usb_network_set_up_dhcpd (usb_moded-network.c:804)
==32357==    by 0x10F5DF: set_dynamic_mode (usb_moded-modesetting.c:446)
==32357==    by 0x10B59F: set_usb_mode (usb_moded.c:379)
==32357==    by 0x10BEE7: msg_handler (usb_moded-dbus.c:108)
==32357==    by 0x48A1567: dbus_connection_dispatch (dbus-connection.c:4658)
==32357==    by 0x48736F7: message_queue_dispatch (dbus-gmain.c:90)
==32357==    by 0x4913B85: g_main_dispatch (gmain.c:3066)
==32357==    by 0x4913B85: g_main_context_dispatch (gmain.c:3642)
==32357==    by 0x4913E11: g_main_context_iterate.isra.5 (gmain.c:3713)
==32357==    by 0x49141D3: g_main_loop_run (gmain.c:3907)
==32357==    by 0x10AF63: main (usb_moded.c:991)
```

### CODE REVIEW

File: src/usb_moded-network.c [hunk 1]
```diff
@@ -299,18 +300,16 @@ static int resolv_conf_dns(struct ipforward_data *ipforward)
...
-   {
-	strcpy(&dest[31], "\0");
-	ret = (strcmp(dest, UDHCP_CONFIG_PATH));
-   }
-   free(dest);
-
-   return(ret);
+  int ret = -1;
+  const size_t bufsiz = 32;
+  char *dest = malloc(bufsiz);
```
Comment (thp) [inline src/usb_moded-network.c]: Why do a heap allocation at all? A stack-allocated buffer of size `PATH_MAX` should do the trick?

```
char dest[PATH_MAX];

size_t len = readlink(UDHCP_CONFIG_LINK, dest, sizeof(dest));

if (len == -1) {
    return -1;
}

dest[len] = 0;
return strcmp(dest, UDHCP_CONFIG_PATH);
```

File: src/usb_moded-network.c [hunk 2]
```diff
@@ -387,20 +386,20 @@ static int write_udhcpd_conf(struct ipforward_data *ipforward, struct mode_list_
...
   {
-	unlink("/etc/udhcpd.conf");
+	unlink(UDHCP_CONFIG_LINK);
   }
   else
 	goto end;
 
 link:
-  symlink(UDHCP_CONFIG_PATH, "/etc/udhcpd.conf");
+  symlink(UDHCP_CONFIG_PATH, UDHCP_CONFIG_LINK);
```
Comment (thp) [inline src/usb_moded-network.c]: Check return value and issue a warning?

File: src/usb_moded-network.c (hunk 1 above)
Comment (monich) [inline src/usb_moded-network.c]: This heap allocation is there for the same reason why it was there in the original code. I don't know why it was there.

If I wrote the original code, I would either allocate a small amount (such as 32 bytes) on stack or a large amount (such as PATH_MAX) from the heap, to reduce the chance of running out of stack space.

When I'm fixing a problem, I am trying to be consistent with the existing code.

Comment (monich) [inline src/usb_moded-network.c]: I think the best is to allocate `sizeof(UDHCP_CONFIG_LINK)`+1 from stack. The code doesn't really need to know the full path pointed by the link, it needs to know whether it equals `UDHCP_CONFIG_LINK` or not.

File: src/usb_moded-network.c (hunk 2 above)
Comment (monich) [inline src/usb_moded-network.c]: That would make sense to me. Will add.

File: src/usb_moded-network.c (hunk 1 above)
Comment (spiiroin) [inline src/usb_moded-network.c]: If you decide to change that, note that sizeof(UDHCP_CONFIG_LINK) == sizeof(char*) -> fail.

Switching from macro to string variable would make the math work, but feels like overkill and requires more brain capacity to grok than some reasonably large buffer from stack.

**Comment by monich** [comment]
Updated per review comments

File: src/usb_moded-network.c (hunk 1 above)
Comment (monich) [inline src/usb_moded-network.c]: @spiiroin: `UDHCP_CONFIG_PATH` (sorry, I meant `UDHCP_CONFIG_PATH`, not `UDHCP_CONFIG_LINK`) is `char[]`, so `sizeof(UDHCP_CONFIG_PATH)` is 27 (length + null terminator)

**Comment by thp** [comment]
LGTM.

**Pull request merged**